### DISCUSSION TITLE
Versioning

### DISCUSSION THREAD

**Discussion by DominicD** [ArthurSonzogni/FTXUI#207] [Ideas]
Hi @ArthurSonzogni  first of all thanks for the great work!

Do you think it would make sense to start Versioning?
I think it would help while using it and also then you can have some release info every time which also helps to know what changed.

**Comment by ArthurSonzogni** [comment]
Yes, that would be great!

What do you recommend exactly? I mean, I can't make a new version with a changelog too often. So maybe a /dev branch and periodically refresh /main with a new version.
The only drawback is this would delay new features by a couple of weeks for the people using /dev

Do you have anything to recommend?

**Reply by DominicD** [reply]
Well I do not have any experience with this myself. However I would suggest having a look at how it is handled in other popular open source projects and think about what style works for you.
Would be great if other user could say what they want as well. Maybe it would make sense to ask the maintainers of the [projects using ftxui](https://github.com/ArthurSonzogni/FTXUI#project-using-ftxui) you linked.

I think [Semantic Versioning](https://semver.org/) makes sense and choosing on of the workflows ([example post](https://www.freshconsulting.com/insights/blog/git-development-workflows-git-flow-vs-github-flow/)).
There is also a lot of great tooling around these e.g. [gitversion](https://github.com/GitTools/GitVersion).

Maybe you already have some experience with it?

**Reply by ezzieyguywuf** [reply]
I don't currently use ftxui, though I've had my eye on it for a while 

Adding releases (ie releasing a version) is a great idea, as it allows package maintainers for various OSes to create a package for your release.

Regarding the actual scheme to use: SemVer is widely used and recognized. v1.0.0 is typically considered a significant initial release. Many open source projects defer v1.0.0 until they've been "out in the wild" for a while and things have stabilized - some never make it there lol.

Releasing versions doesn't have to be a big headache. I encourage you to do so with as little effort as possible and then scale as (or if) needed.

**Comment by DominicD** [comment]
Thank you for creating a first release and the extensive changelog!
One more suggestion I have now that I see it, is to actually use the Github Release feature. 
If you use it it creates the tag and also gives a nicer view. Here you can see an explanation: [releases vs tags](http://practicalseries.com/1002-vcs/09-06-github.html#js--090600)

Of course its up to you if you care about it. I am happy anyway 👍

**Comment by ArthurSonzogni** [comment]
Github release is now automatically used whenever I push a new tag. Moreover, artefact are automatically created (Deb, RPM, etc...).

@DominicD, would you have more suggestions after that?

**Reply by DominicD** [reply]
Perfect. Only suggestion is keep up the great work :)

**Comment by BurningEnlightenment** [comment]
@ArthurSonzogni it's been a while since 3.0.0 has been released, are there any plans for "officially" releasing the current changes?

**Reply by ArthurSonzogni** [reply]
I feel the need to switch version.
There is still one more breaking change I would like to add:
https://github.com/ArthurSonzogni/FTXUI/tree/textarea

After that, it shouldn't take too long.

**Reply by BurningEnlightenment** [reply]
That's hot stuff, thanks!
Also, it would be nice if critical/major bug fixes like #421 were "backported" and packaged as early/seperate patch releases.

**Reply by ArthurSonzogni** [reply]
I ended up releasing v4.0.0 earlier. Mainly because of the clang bug affecting v3.0.0.
